### ISSUE TITLE
 High Memory usage with my trackee channels

### ISSUE THREAD

**Issue by shoaibahmedqureshi** [ably/ably-ios#691]
We are experiencing high memory usage currently with our application channels and we are not sure about the reason and sometimes it takes more than a GB which is not good for a mobile device . I have simulated this issue with our own channels in a separate project and it has been shared with @funkyboy as project contains our channels so we are not keeping it public on GitHub repository.
We suspect that https://github.com/ably/ably-ios/issues/673  could be because of this issue.
When memory goes high often we get the message "Count response is greater then the total of pending message."
Feel free to further communicate for this project. 
Following is the project url https://drive.google.com/drive/folders/1oYMa0wakNKXfSH7M8gAzPX_2EnShcdTH

<img width="1322" alt="message" src="https://user-images.githubusercontent.com/7335486/36232134-652b9acc-1202-11e8-918b-325aec780061.png">

<img width="1369" alt="screen shot 2018-02-15 at 3 19 23 am" src="https://user-images.githubusercontent.com/7335486/36233250-d335bdd2-1206-11e8-8ca2-f0e72834890f.png">

![Uploading Screen Shot 2018-02-15 at 1.56.50 AM.png…]()

Screen Shot 2018-02-15 at 1.56.50 AM

**Comment by funkyboy** [comment]
@shoaibahmedqureshi thanks for reporting this. I asked access to the project URL. Once granted I'll take a look.

**Comment by funkyboy** [comment]
@shoaibahmedqureshi I saw you sent a Google Drive link (above) and then shared some other link via email. Please let me know which project I should check out. 
Also, since I'll debug with data that belong to your channels, I am gonna need the API key that you used to reproduce this. You have my email.

**Comment by shoaibahmedqureshi** [comment]
@funkyboy both point to the same code and if they are not please let me know and also I am emailing you the API Key .

**Comment by funkyboy** [comment]
@shoaibahmedqureshi received. What are the steps to reproduce a 1GB memory situation?

**Comment by shoaibahmedqureshi** [comment]
@funkyboy just run the code and wait you will see memory increasing if channels are subscribed and messages are coming and if it does not reproduce then try 2 3 times and it will reproduce.

**Comment by funkyboy** [comment]
@shoaibahmedqureshi The memory grows because the `tick` function is called every .3 seconds to send fake data to the channels. I put it there initially just to simulate a quite heavy workload. Hence the `messages` array will grow indefinitely. I don't think it's a real use case :)

**Comment by hrubbani** [comment]
@funkyboy 
where is this tick() function that sends the data ? is it in your sample project code ? 

we are experiencing this issue in live app, when  it is start listening all demobus channels, memory starting increasing and after sometime app hangs. App doesn't keep data, it just moves marker on map with every update.

memory starts keep increasing when we tried those channels in your provided project. On these channels data comes after every second or so.

**Comment by funkyboy** [comment]
@hrubbani It's in the project you sent me. See screenshot.
![screen shot 2018-02-16 at 12 47 10](https://user-images.githubusercontent.com/348355/36306343-97d6b810-1317-11e8-9c9d-7c1c53f6fdba.png)

**Comment by hrubbani** [comment]
thanks @funkyboy 
@shoaibahmedqureshi can you remove this code and try again to see if that keeps down the memory ? or clarify in case if this is not relevant after your changes

**Comment by shoaibahmedqureshi** [comment]
@funkyboy I removed tick function and introduced stateChange method which tracks channel state change and reattaches any failed or detached channels and I can see memory going up . I shared the link with you please let me know if you did not receive it.

**Comment by funkyboy** [comment]
@shoaibahmedqureshi I received it.
I could not get the app to reach 1GB of RAM consumption, but after running for 1h20m I got it to ~650MB. Yet I don't understand the concern.

Every new message (regardless of the channel) is stored in `messages`, line 67. That is an in-memory array that will grow as new messages are received. Thus RAM consumption will grow. 
I don't fully know the scope of your app, but it's likely that:
- you'll store messages somewhere else (Core Data, file)
- you'll get rid of old messages
- some other custom logic

That's what I mean when I say that this project, as it is implemented right now, is not a "real use case".
Or am I misunderstanding your report?

Nevertheless, I'll test this again next week - without storing data in `messages` - to see what happens.

**Comment by shoaibahmedqureshi** [comment]
@funkyboy I am wondering why it is not reproducing in your case and even if you comment out the code to append in the message array. It often goes in GBs , I am attaching an screenshot which I just reproduced the issue and code to append message array is also commented out. We suspect that after receiving messages memory is somehow not releasing and it happens before we save messages in memory . please refer to screenshot below.

<img width="1430" alt="screen shot 2018-02-17 at 1 45 45 am" src="https://user-images.githubusercontent.com/7335486/36328685-fa39ae46-1384-11e8-9047-e35b1c4e3268.png">

**Comment by shoaibahmedqureshi** [comment]
Also please refer to screen shot below.
![Uploading Screen Shot 2018-02-17 at 1.57.14 AM.png…]()

**Comment by shoaibahmedqureshi** [comment]
Please also refer to screen shot below.
<img width="1433" alt="screen shot 2018-02-17 at 1 57 14 am" src="https://user-images.githubusercontent.com/7335486/36329076-6f25ea34-1386-11e8-8bd1-928ab01d0d28.png">

**Comment by funkyboy** [comment]
@shoaibahmedqureshi ok, let's see how else we can tackle this.

First I'd make sure to we run the app on the same Xcode version and simulator.
I am running Xcode `9.2 (9C40b)` and I tested it on a `iOS 11.2 iPhone 8 Plus` simulator.
I think you are using the same simulator, but is the version of Xcode the same? If not, any chance you can test it with the 9.2 version?

Also, is the content on channels "constantly updating"? If not, we need to coordinate about this.

**Comment by funkyboy** [comment]
@shoaibahmedqureshi 
I removed the `startPublishing` function and I see zero traffic on the channels.

**Comment by shoaibahmedqureshi** [comment]
@funkyboy i checked on 9.0 and 9.2 both and i am able to reproduce this . Please  make sure if you are not receiving data , you are using latest project i gave  you and latest one contains channelStateChange method and even then if you have trouble reproducing it we can jump on call.
@hrubbani can you please clarify his point "is the content on channels "constantly updating"?" 

<img width="1156" alt="screen shot 2018-02-19 at 6 05 22 pm" src="https://user-images.githubusercontent.com/7335486/36379412-4b48fa2c-15a0-11e8-9841-a2500be49773.png">

**Comment by funkyboy** [comment]
@shoaibahmedqureshi you sent me 2-3 projects and I don't understand anymore which one you'd like me to run. I am online on Skype. Please send it over there.

**Comment by shoaibahmedqureshi** [comment]
@funkyboy I gave you two projects the latest one is AblyTest-New.zip. I am on available on skype .

**Comment by funkyboy** [comment]
For future reference: tap the refresh button to trigger the reception of messages.

**Comment by shoaibahmedqureshi** [comment]
@funkyboy somehow I am unable to send video over skype so let me try and upload here.

**Comment by shoaibahmedqureshi** [comment]
let me try and upload it somewhere and send you the link as can't upload it here either due to format issue.

**Comment by funkyboy** [comment]
For future reference video is here: https://www.youtube.com/watch?v=PSBPgbd6_UQ

**Comment by funkyboy** [comment]
Further inspection suggests that setting a low log level like:

`options?.logLevel = .warn` or `options!.logLevel = .none`

brings the memory consumption down quite a lot.
@shoaibahmedqureshi will confirm this with further tests.

**Comment by shoaibahmedqureshi** [comment]
@funkyboy I performed further tests on the same project given by you to me over skype in yesterday's session in which we performed tests using different log levels and further inspection suggests that keeping log level like options!.logLevel = .none does not keep the issue away from reproducing please refer to the video https://www.youtube.com/watch?v=zn0yQIkmQW4 .
Also notice in the video that Network shows few KBs while memory grows in several MBs and as per @hrubbani our message is of few bytes only. Clearly in the project we are not saving message anywhere . The video shows that memory only came down significantly when channels failed and went up again upon refresh. Please  investigate further as the test in the video suggests that even changing level does not make memory stable.

**Comment by funkyboy** [comment]
@shoaibahmedqureshi thanks for your report. We are going to take a look at this. As it's difficult to replicate, it might take a while.

**Comment by hrubbani** [comment]
@funkyboy  It will be great if we could get update/resolution on this as it is consistently being produced for us, and causing un-responsiveness/crashes of live application. Let us know if you need any further assistance to produce this issue.

**Comment by funkyboy** [comment]
@hrubbani I have been able to reproduce it only once. Still investigating the cause. I'll let you know as soon as I have an update.

**Comment by funkyboy** [comment]
@shoaibahmedqureshi Please try the following:

- In the project folder run `pod deintegrate`
- delete `AblyTest.xcworkspace` and `Podfile.lock`
- change the `Podfile` like [this](https://gist.github.com/funkyboy/e0ba3ddbfe48280563eea03c86177c74)
- run `pod install`
- open the workspace and perform your test again

Let me know. If this fixes it, we can proceed and make a new release.

**Comment by shoaibahmedqureshi** [comment]
@funkyboy thanks for your response . I tried above mentioned steps . I can see little bit of improvement however in some cases I noticed if i leave application on for an hour or so and messages keep coming , the memory goes high.

**Comment by funkyboy** [comment]
@shoaibahmedqureshi good. I'll try that. 
To be clear, what do you mean by "the memory goes high"? 1GB? More? And stays there always? Or drops after a while?

**Comment by funkyboy** [comment]
@shoaibahmedqureshi Update: there's an internal method getting called many times per second that impacts on memory consumption. I am investigating that now.

**Comment by shoaibahmedqureshi** [comment]
@funkyboy Thanks for your update. yes I noticed it going to 1.05 GB and after that It did not drop.

**Comment by funkyboy** [comment]
@shoaibahmedqureshi I pushed one more change. 
Please try the following:

- In the project folder run `pod deintegrate`
- delete `AblyTest.xcworkspace` and `Podfile.lock`
- change the `Podfile` like [this](https://gist.github.com/funkyboy/e0ba3ddbfe48280563eea03c86177c74)
- run `pod install`
- open the workspace and perform your test again

Let me know how it goes.

**Comment by funkyboy** [comment]
@shoaibahmedqureshi any news about this?

**Comment by hrubbani** [comment]
@funkyboy Shoaib tested this today and found that there is some improvement, like, memory is not going too high (1 GB and more) very fast. it goes up but takes more time  (an hour or more). but when it takes around 300 MB application stops responding. For me it looks it is not releasing memory because even 300MB is a lot  in 20-30 minutes. our single message is less than 300 bytes, so in no circumstances it should be taking 300 MB. also it should keep releasing memory as messages are delivered to app.  All these findings are when Shoaib tested code directly into our app.

In your testing with your test project, how much maximum memory you have seen when testing with our demo channels ?

**Comment by funkyboy** [comment]
@hrubbani In my testing (in the simulator) the average memory consumption is around 50MB.

@shoaibahmedqureshi Can you please run the following in the Terminal on your machine?
`curl -I http://rest.ably.io/time` and tell me the value of `X-Ably-Serverid: `?

Should be something of this form: `frontend.1937.1.eu-west-1-A.i-0634f88763a4f91bc`
It's public data so you can post it here.

**Comment by shoaibahmedqureshi** [comment]
@funkyboy here it is X-Ably-Serverid: frontend.6458.2.eu-west-1-A.i-0f7124ce380adf4c5

**Comment by funkyboy** [comment]
@shoaibahmedqureshi One more thing. Can you delete this code?
```
case .failed?:     
  channel?.subscribe() { message in

  }
  break
```
When a channel is in a failed state, trying to subscribe is pointless. It is gonna error for sure.

**Comment by shoaibahmedqureshi** [comment]
@funkyboy done.

**Comment by funkyboy** [comment]
@shoaibahmedqureshi I sent you a new project via email. Can you please test it and let me know?

**Comment by shoaibahmedqureshi** [comment]
@funkyboy I had tested it earlier after removing
 case .failed?:     
  channel?.subscribe() { message in

  }
  break 
and it turns out that memory hike process was not as quick as it used to be but with each message coming it goes little bit up and never comes down and  it started from 50 MB  and after sometime (approx and hour or so ) It reached 300 MB and I did not witness any drop during this process and with this continued hike it might cross 1 GB in some hours What I am noticing is that once message is received and in test project it is not being used any where the memory is  . To reproduce you need run application and leave it running in foreground and ensure that messages keep coming. With that latest project you sent me why have you commented out

**Comment by funkyboy** [comment]
@shoaibahmedqureshi 
> With that latest project you sent me why have you commented out the rest of channels except one channe ?

That was for testing. Please reactivate them all, so we can test it the same condition. 

The last project I sent you via email should use even less memory. Let me know.

**Comment by shoaibahmedqureshi** [comment]
@funkyboy I will test it and let you know . I see there is no way set log levels in that project and all ?  is it pointing to some special commit/branch ? if so can you provide us the pod link as well so we can test it in our actual project as well ?

**Comment by funkyboy** [comment]
@shoaibahmedqureshi that’s exactly the point of the project I sent you. Remove any kind of logging and see the impact. 

At the moment I don’t have yet a branch that excludes logging.

**Comment by shoaibahmedqureshi** [comment]
@funkyboy I tested your test project given by you and memory seemed alot improved and under control and if you could please create a pod link for it . I would further test it in our actual application.

**Comment by funkyboy** [comment]
@shoaibahmedqureshi the "workaround" is to comment out the content of [this function](https://github.com/ably/ably-ios/blob/master/Source/ARTSentry.m#L430) which you'll find in the `Pods/Ably/Source` folder of your original project.

I'd like to point out that memory tests so far have been performed in debug mode, which does NOT reflect the behaviour of an app in production. 

I'll soon post a public project that we can use as a test bed to measure memory consumption.

**Comment by funkyboy** [comment]
@shoaibahmedqureshi as agreed via email I uploaded the [sample project](https://github.com/funkyboy/AblyMemoryTest), but anonymized. 
See the README file for instructions on measuring memory consumption.

I ran a bunch of tests on that project but I could NOT reproduce an ever growing memory consumption.
Here's a screenshot for reference. Average consumption was around 20MB.

![screen shot 2018-03-01 at 14 05 48](https://user-images.githubusercontent.com/348355/36846726-11db4fd8-1d5c-11e8-8ee2-7e3584774334.png)

Feel free to branch the project and push changes that you think are relevant to reproduce the issue.

**Comment by funkyboy** [comment]
@shoaibahmedqureshi While testing the sample project on a device using a 3G connection I noticed some excessive memory consumption (although not ever growing). So I [tweaked the logger](https://github.com/ably/ably-ios/commit/62df8ec35c5e1a1c83d00b1dd4a6272f5d72eee4). 

Please follow the [Installation steps in the README file](https://github.com/funkyboy/AblyMemoryTest/blob/master/README.md#installation) to pull the new branch and run new tests.

Let me know.

**Comment by shoaibahmedqureshi** [comment]
@funkyboy With the workaround (commenting out method) kept memory quite low and in control , yet the testbed you shared and  by pulling  tweaked logger fix (without commenting out method)  I could see it going above 1 GB and and I had set the log level .none. I did not use profiler yet. I tested it on simulator and @hrubbani  is further testing (tweaked logger fix) on device in our actual app. Just add the key and add all those 9 channels with had in previous test project and the only change I introduced was channelStateChange method which too is there in previous project and even I was able to reproduce it without introducing this method too ,but I am still committing it for further clarity.  I was able to reproduce above 1 GB memory 3 to 4 times.

<img width="1432" alt="screen shot 2018-03-06 at 5 57 08 am" src="https://user-images.githubusercontent.com/7335486/37008477-02fa25e2-2104-11e8-9c2e-5b0008df3f1b.png">

**Comment by shoaibahmedqureshi** [comment]
For some reason I am unable to push the code but the method is there in the previous project and I was able to reproduce it without that as well.

**Comment by funkyboy** [comment]
@shoaibahmedqureshi If you are referring to [this project](https://github.com/funkyboy/AblyMemoryTest) you can't push to master but you can create a branch and a PR so I can see the diff.

**Comment by funkyboy** [comment]
@shoaibahmedqureshi @hrubbani while inspecting another issue I found a bug in our logging code.
I pushed [an update](https://github.com/ably/ably-ios/tree/fix-memory-consumption-bug).

Please retry testing the [test bed project](https://github.com/funkyboy/AblyMemoryTest) by following the instructions in the README.

Feel free also to [pull the branch](https://github.com/funkyboy/AblyMemoryTest/blob/master/Podfile#L6) into your original project and see how it goes.

**Comment by shoaibahmedqureshi** [comment]
@funkyboy from the initial testing the memory looks improved on log level .none , however we are further testing it and will share the details in case of any issues.

**Comment by funkyboy** [comment]
@shoaibahmedqureshi thanks for the update. I invite you to test also the `.error` level, which is presumably the value you'll use in production.

**Comment by funkyboy** [comment]
@shoaibahmedqureshi any news on this? We will make a new release soon with [these changes](https://github.com/ably/ably-ios/pull/702)

**Comment by hrubbani** [comment]
@funkyboy  so far so good, app looks much better after the latest changes, no more hangs, or freezes. We will let you know if we found any thing unusual in our testing. For now, these can be released, from our side.

**Comment by AndrewBarba** [comment]
Any chance this gets merged today? We just got hit really hard with this, thousands of our users experienced crashing during a live event we just ran

**Comment by mattheworiordan** [comment]
> Any chance this gets merged today? We just got hit really hard with this, thousands of our users experienced crashing during a live event we just ran

@AndrewBarba how do you know this issue is related to the problems you are having?  Can you provide some more detail on what problem you are having?

**Comment by AndrewBarba** [comment]
Because of the OOM crashes I saw in Fabric during our live event

**Comment by AndrewBarba** [comment]
@mattheworiordan @funkyboy @hrubbani 

Here's memory usage for current sdk:

<img width="806" alt="screen shot 2018-03-15 at 3 54 31 pm" src="https://user-images.githubusercontent.com/1316152/37488010-bb3e09cc-2869-11e8-9a05-70f6cfcc6add.png">

<img width="995" alt="screen shot 2018-03-15 at 3 54 46 pm" src="https://user-images.githubusercontent.com/1316152/37488016-c1799ae0-2869-11e8-97b5-1c58f242e255.png">

**Comment by AndrewBarba** [comment]
`fix-memory-consumption` branch:

<img width="1035" alt="screen shot 2018-03-15 at 4 00 25 pm" src="https://user-images.githubusercontent.com/1316152/37488139-0e321b32-286a-11e8-9069-377ae63a7bc8.png">

<img width="985" alt="screen shot 2018-03-15 at 4 00 30 pm" src="https://user-images.githubusercontent.com/1316152/37488141-11f304a2-286a-11e8-99df-aa982f39cdb5.png">

**Comment by funkyboy** [comment]
@shoaibahmedqureshi @AndrewBarba @hrubbani This is now fixed in `1.0.12`

**Comment by AndrewBarba** [comment]
Thanks!
